jobmanager/AsyncWaiter.h

```diff
@@ -0,0 +1,195 @@
+        startTime(startTime),
+        endTime(endTime)
+    { }
+
+    std::unique_ptr<AsyncWaitTask> task;
+    uint64_t startTime;
+    uint64_t endTime;
+};
+
+struct CheckJob : CountedJob
+{
+    CheckJob(AsyncWaiter& asyncWaiter)
+        : CountedJob(asyncWaiter._jobCounter),
+          _asyncWaiter(asyncWaiter)
+    { }
+
+    void run() final { _asyncWaiter.runCheck(); }
+
+    AsyncWaiter& _asyncWaiter;
+};
+
+struct ScheduleCheckJobTimer : CountedJob
+{
+    ScheduleCheckJobTimer(AsyncWaiter& asyncWaiter)
+        : CountedJob(asyncWaiter._jobCounter),
+          _asyncWaiter(asyncWaiter)
+    { }
+
+    void run() final { _asyncWaiter._jobQueue.template addJob<CheckJob>(_asyncWaiter); }
+
+    AsyncWaiter& _asyncWaiter;
+};
+
+struct AddTaskJob : CountedJob
+{
+    template<class...TaskArgs>
+    AddTaskJob(AsyncWaiter& asyncWaiter, uint64_t timeout, std::unique_ptr<AsyncWaitTask>&& task)
+        : CountedJob(asyncWaiter._jobCounter),
+         _asyncWaiter(asyncWaiter),
+          _task(std::move(task)),
+          _timeout(timeout)
+    { }
+
+    void run() final { _asyncWaiter.addNewTaskInternal(_timeout, std::move(_task)); }
+
+    AsyncWaiter& _asyncWaiter;
+    std::unique_ptr<AsyncWaitTask> _task;
+    uint64_t _timeout;
+};
+
+public:
+    explicit AsyncWaiter(JobManager& jobManager, size_t maxCheckInterval, size_t jobQueueSize = 128);
+
+    template<class Task, class... Args>
+    void emplaceTask(uint64_t timeout, Args&&...args);
+    template<class Task>
+    void addTask(uint64_t timeout, Task&& task);
+
+    bool hasTaskWaiting() const { return _jobCounter.load() != 0; }
+
+private:
+    void runCheck();
+    void scheduleNexCheck(uint64_t timeoutTs);
+    void addNewTaskInternal(uint64_t timeout, std::unique_ptr<AsyncWaitTask>&& task);
+    bool checkTaskCompletion(AsyncWaitTask& task);
+
+private:
+    JobQueue _jobQueue;
+    std::vector<AsyncEntry> _tasks;
+    std::atomic_uint32_t _jobCounter;
+    uint64_t _maxCheckInterval;
+    uint64_t _timerId;
+};
+
+
+inline AsyncWaiter::AsyncWaiter(JobManager& jobManager, size_t maxCheckInterval, size_t jobQueueSize)
+    : _jobQueue(jobManager, jobQueueSize),
+      _tasks(),
+      _jobCounter(0),
+      _maxCheckInterval(maxCheckInterval),
+      _timerId(reinterpret_cast<uint64_t>(this))
+{
+}
+
+inline void AsyncWaiter::runCheck()
+{
+    const auto timestamp = utils::Time::getAbsoluteTime();
+    auto itEnd = _tasks.end();
+    auto nextTimeout = _maxCheckInterval;
+    for (auto it = _tasks.begin(); it != itEnd; )
+    {
+        const bool isCompleted = checkTaskCompletion(*it->task);
+        const bool isEnded = isCompleted || timestamp >= it->endTime;
+        if (isEnded)
+        {
+            if (!isCompleted)
+            {
+                it->task->onTimeout();
+            }
+
+            *it = std::move(*(itEnd - 1));
+            --itEnd;
+        }
+        else
+        {
+            nextTimeout = std::min(nextTimeout, it->endTime - timestamp);
```

Comment:
You will miss setting next check 0ms if the timestamp passed endtime. endtime - timestamp will become huge and we will wait _maxCheckInterval, unless there is another task with slightly less time left.

Reply:
Here `it->endTime` is alway bigger than `timestamp`, otherwise the task ended by timeout